// playwright-core/lib/rpc/server/downloadDispatcher.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the 'License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DownloadDispatcher = void 0;
const dispatcher_1 = require("./dispatcher");
class DownloadDispatcher extends dispatcher_1.Dispatcher {
    constructor(scope, download) {
        super(scope, download, 'download', {
            url: download.url(),
            suggestedFilename: download.suggestedFilename(),
        });
    }
    async path() {
        return this._object.path();
    }
    async failure() {
        return this._object.failure();
    }
    async delete() {
        await this._object.delete();
    }
}
exports.DownloadDispatcher = DownloadDispatcher;
