// playwright-core/lib/rpc/client/browser.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Browser = void 0;
const browserContext_1 = require("./browserContext");
const channelOwner_1 = require("./channelOwner");
const events_1 = require("../../events");
class Browser extends channelOwner_1.ChannelOwner {
    constructor(scope, guid, initializer) {
        super(scope, guid, initializer, true);
        this._contexts = new Set();
        this._isConnected = true;
        this._isClosedOrClosing = false;
        this._channel.on('close', () => {
            this._isConnected = false;
            this.emit(events_1.Events.Browser.Disconnected);
            this._isClosedOrClosing = true;
            this._scope.dispose();
        });
    }
    static from(browser) {
        return browser._object;
    }
    static fromNullable(browser) {
        return browser ? Browser.from(browser) : null;
    }
    async newContext(options = {}) {
        delete options.logger;
        const context = browserContext_1.BrowserContext.from(await this._channel.newContext(options));
        this._contexts.add(context);
        context._browser = this;
        return context;
    }
    contexts() {
        return [...this._contexts];
    }
    async newPage(options = {}) {
        delete options.logger;
        const context = await this.newContext(options);
        const page = await context.newPage();
        page._ownedContext = context;
        context._ownerPage = page;
        return page;
    }
    isConnected() {
        return this._isConnected;
    }
    async close() {
        if (this._isClosedOrClosing)
            return;
        this._isClosedOrClosing = true;
        await this._channel.close();
    }
}
exports.Browser = Browser;
