// playwright-core/lib/rpc/client/download.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Download = void 0;
const fs = require("fs");
const channelOwner_1 = require("./channelOwner");
class Download extends channelOwner_1.ChannelOwner {
    static from(download) {
        return download._object;
    }
    constructor(scope, guid, initializer) {
        super(scope, guid, initializer);
    }
    url() {
        return this._initializer.url;
    }
    suggestedFilename() {
        return this._initializer.suggestedFilename;
    }
    async path() {
        return this._channel.path();
    }
    async failure() {
        return this._channel.failure();
    }
    async createReadStream() {
        const fileName = await this.path();
        return fileName ? fs.createReadStream(fileName) : null;
    }
    async delete() {
        return this._channel.delete();
    }
}
exports.Download = Download;
